src/socket/tcp.rs

```diff
@@ -1110,9 +1122,36 @@ impl<'a> TcpSocket<'a> {
         }
 
         if let Some(ack_number) = repr.ack_number {
+            let old_local_seq_no = self.local_seq_no;
             // We've processed everything in the incoming segment, so advance the local
             // sequence number past it.
             self.local_seq_no = ack_number;
+            // During retransmission, a higher sequence number can get acknowledged than
+            // the last calculated after retransmitting an older packet. Thus forward
+            // to the local sequence number. In normal operation, nothing should be done here.
+            // Due to wrapping of the sequence numbers it is not enough to check
+            // the two cases of local_seq_no < remote_last_seq to see if it is a retransmission
+            // but the two wrapped cases also have to be considered. To distinguish them
+            // old_local_seq_no is needed.
+            if self.local_seq_no < self.remote_last_seq &&
+               !(self.local_seq_no < old_local_seq_no && old_local_seq_no < self.remote_last_seq) {
+                // This is the simple case for normal operation.
+            } else if self.remote_last_seq < old_local_seq_no && old_local_seq_no < self.local_seq_no {
+                // Second case for normal operation where remote_last_seq is already wrapped.
+            } else if (self.remote_last_seq < self.local_seq_no &&
+                       !(self.remote_last_seq < old_local_seq_no && old_local_seq_no < self.local_seq_no)) ||
+                       // Like the second case but old_loc_seq_no is at a different position.
+                       // Next like the first case but again because local_seq_no is wrapped,
+                       // but old_loc_seq_no is at a different position.
+                      (self.local_seq_no < old_local_seq_no && old_local_seq_no < self.remote_last_seq) {
+                // These two cases describe retransmission where remote_last_seq lags behind local_seq_no.
+                // The value for the next sequence number to be used is updated.
+                self.remote_last_seq = self.local_seq_no;
```

Comment:
I can remove all empty cases and the unreachable which I added to see if something was missed. Then if would just be the following:

```
if (self.remote_last_seq < self.local_seq_no &&
   !(self.remote_last_seq < old_local_seq_no && old_local_seq_no < self.local_seq_no)) ||
  (self.local_seq_no < old_local_seq_no && old_local_seq_no < self.remote_last_seq) {
    self.remote_last_seq = self.local_seq_no;
}
```

Reply:
I am sure that you have discovered a real issue, but I do not seem to understand your explanation. Do you think you could expand on it or rephrase it in a different way?